Source/santa-driver/SantaDriverClient.cc

```diff
@@ -110,97 +110,78 @@ IOReturn SantaDriverClient::clientMemoryForType(
 
 #pragma mark Callable Methods
 
-IOReturn SantaDriverClient::open() {
-  if (isInactive()) return kIOReturnNotAttached;
+IOReturn SantaDriverClient::open(
+    OSObject *target,
+    void *reference,
+    IOExternalMethodArguments *arguments) {
+  SantaDriverClient *me = OSDynamicCast(SantaDriverClient, target);
+  if (!me) return kIOReturnBadArgument;
 
-  if (!myProvider->open(this)) {
+  if (me->isInactive()) return kIOReturnNotAttached;
+  if (!me->myProvider->open(me)) {
     LOGW("A second client tried to connect.");
     return kIOReturnExclusiveAccess;
   }
 
   return kIOReturnSuccess;
 }
 
-IOReturn SantaDriverClient::static_open(
-    SantaDriverClient *target,
-    void *reference,
-    IOExternalMethodArguments *arguments) {
-  if (!target) return kIOReturnBadArgument;
-  return target->open();
-}
+IOReturn SantaDriverClient::allow_binary(
+    OSObject *target, void *reference, IOExternalMethodArguments *arguments) {
+  SantaDriverClient *me = OSDynamicCast(SantaDriverClient, target);
+  if (!me) return kIOReturnBadArgument;
+  if (arguments->scalarInput == nullptr) return kIOReturnBadArgument;
```

Comment:
If the userland caller of IOConnectCallScalarMethod promises an `input` with `inputCnt` but does not give one, the caller will crash.

```
Thread 11 Crashed:: Dispatch queue: com.apple.NSXPCConnection.user.15028
0   libsystem_platform.dylib           	0x00007fff986d9fbd _platform_memmove$VARIANT$Haswell + 157
1   libsystem_c.dylib                  	0x00007fff913d7f13 __memcpy_chk + 22
2   com.apple.framework.IOKit          	0x00007fff915b3b19 io_connect_method + 120
3   com.apple.framework.IOKit          	0x00007fff915543c4 IOConnectCallScalarMethod + 111
```

If the userland caller provides no `inputCnt` promise and no `input`, the decision threads are never woken up and I experience deadlock. We may want to look into other options for preventing deadlock if the vnode_ids don't make the round trip or are corrupted.